pkg/cdi/cdi.go

```diff
@@ -0,0 +1,124 @@
+	cdiSpecs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	"github.com/golang/glog"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types"
+)
+
+const cdiSpecPrefix = "sriov-dp-"
+
+// CDI represents CDI API required by Device plugin
+type CDI interface {
+	CreateCDISpecForPool(resourcePrefix string, rPool types.ResourcePool) error
+	CreateContainerAnnotations(devicesIDs []string, resourcePrefix, resourceKind string) (map[string]string, error)
+	CleanupSpecs() error
+}
+
+// Impl implements CDI interface
+type Impl struct {
+}
+
+// CreateCDISpecForPool creates CDI spec file with specified devices
+func (c *Impl) CreateCDISpecForPool(resourcePrefix string, rPool types.ResourcePool) error {
+	err := c.CleanupSpecs()
+	if err != nil {
+		glog.Error("can not cleanup old spec files", err)
+		return err
+	}
+	cdiDevices := make([]cdiSpecs.Device, 0)
+	cdiSpec := cdiSpecs.Spec{
+		Version: cdiSpecs.CurrentVersion,
+		Kind:    resourcePrefix + "/" + rPool.GetCDIName(),
+		Devices: cdiDevices,
+	}
+
+	for _, dev := range rPool.GetDevices() {
+		containerEdit := cdiSpecs.ContainerEdits{
+			DeviceNodes: make([]*cdiSpecs.DeviceNode, 0),
+		}
+
+		for _, spec := range rPool.GetDeviceSpecs([]string{dev.GetID()}) {
+			deviceNode := cdiSpecs.DeviceNode{
+				Path:        spec.ContainerPath,
+				HostPath:    spec.HostPath,
+				Permissions: "rw",
+			}
+			containerEdit.DeviceNodes = append(containerEdit.DeviceNodes, &deviceNode)
+		}
+		device := cdiSpecs.Device{
+			Name:           dev.GetID(),
+			ContainerEdits: containerEdit,
+		}
+		cdiSpec.Devices = append(cdiSpec.Devices, device)
+	}
+
+	err = cdi.GetRegistry().SpecDB().WriteSpec(&cdiSpec, cdiSpecPrefix+resourcePrefix)
+	if err != nil {
+		glog.Errorf("CreateCDISpecForPool(): can not create CDI json: %q", err)
+		return err
+	}
+
+	return nil
+}
+
+// CreateContainerAnnotations creates container annotations based on CDI spec for a container runtime
+func (c *Impl) CreateContainerAnnotations(devicesIDs []string, resourcePrefix, resourceKind string) (map[string]string, error) {
+	annotations := make(map[string]string, 0)
+	annoKey, err := cdi.AnnotationKey(resourcePrefix, resourceKind)
+	if err != nil {
+		glog.Error("can not create container annotation", err)
+		return nil, err
+	}
+	devices := make([]string, 0)
+	for _, id := range devicesIDs {
+		devices = append(devices, cdi.QualifiedName(resourcePrefix, resourceKind, id))
+	}
+	annoValue, err := cdi.AnnotationValue(devices)
+	if err != nil {
+		glog.Error("can not create container annotation", err)
+		return nil, err
+	}
+	annotations[annoKey] = annoValue
+
+	return annotations, nil
+}
+
+// CleanupSpecs removes previously-created CDI specs
+func (c *Impl) CleanupSpecs() error {
+	for _, dir := range cdi.GetRegistry().GetSpecDirectories() {
+		specs, err := filepath.Glob(filepath.Join(dir, cdiSpecPrefix+"*"))
+		if err != nil {
```

Comment:
will this one fail if dir doesnt exist ?

if yes, then for such error we need to skip.

Reply:
> Glob returns the names of all files matching pattern or nil if there is no matching file. The syntax of patterns is the same as in Match. The pattern may describe hierarchical names such as /usr/*/bin/ed (assuming the Separator is '/').
> Glob ignores file system errors such as I/O errors reading directories. The only possible returned error is ErrBadPattern, when pattern is malformed.

https://pkg.go.dev/path/filepath#Glob